Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp

```diff
@@ -164,4 +176,82 @@ ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(std::span<
 
     return compressedData.releaseNonNull();
 }
+
+
+#if PLATFORM(COCOA)
+ExceptionOr<bool> CompressionStreamEncoder::initializeAppleCompressionFramework() 
+{
+    m_initialized = true;
+    m_usingAppleCompressionFramework = true;
+
+    auto result = compression_stream_init(&m_stream, COMPRESSION_STREAM_ENCODE, COMPRESSION_BROTLI);
+    if (result != COMPRESSION_STATUS_OK)
+        return Exception { ExceptionCode::TypeError, "Initialization Failed."_s };
+
+    return true;
+}
+
+ExceptionOr<Ref<JSC::ArrayBuffer>> CompressionStreamEncoder::compressAppleCompressionFramework(std::span<const uint8_t> input)
+{
+    m_didFinish = true;
+
+    size_t allocateSize = std::max(input.size(), startingAllocationSize);
+    auto storage = SharedBufferBuilder();
+
+    compression_status result;    
+    bool shouldDecompress = true;
+
+    if (!m_initialized) {
+        auto initializeResult = initializeAppleCompressionFramework();
+        if (initializeResult.hasException())
+            return initializeResult.releaseException();
+    }
+
+    m_stream.src_ptr = input.data();
+    m_stream.src_size = input.size();
+    
+    while (shouldDecompress) {
+        Vector<uint8_t> output;
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
+
+            if (allocateSize < startingAllocationSize)
+                return Exception { ExceptionCode::OutOfMemoryError };
+
+            continue;
+        }
+
+        output.grow(allocateSize);
+
+        m_stream.dst_ptr = output.data();
+        m_stream.dst_size = output.size();
+
+        result = compression_stream_process(&m_stream, m_didFinish ? COMPRESSION_STREAM_FINALIZE : 0);
+        
+        if (result == COMPRESSION_STATUS_ERROR)
+            return Exception { ExceptionCode::TypeError, "Failed to Encode Data."_s };
+
+        if ((result == COMPRESSION_STATUS_END && m_stream.src_size)
+            || (m_didFinish && m_stream.src_size))
+            return Exception { ExceptionCode::TypeError, "Extra bytes past the end."_s };
```

Comment:
I would expect this to result in retrying with a larger buffer then an OutOfMemoryError if it can't be allocated.

Reply:
I think we may be able to remove this check from the compression section as this was designed to handle extraneous bytes added to the end of a compressed message. This check I believe is mainly relevant to decompression.